appscale/agents/ec2_agent.py

```diff
@@ -673,60 +681,124 @@ def terminate_instances(self, parameters):
-    # Sending a second terminate to a terminated instance to remove it
-    # from the system (ie no more in describe-instances).  This helps when
-    # bringing deployments up and down frequently and instances are still
-    # associated with keyname (although they are terminated).
+    status_filters = {"instance-state-name": 'terminated',
+                      "key-name": parameters[self.PARAM_KEYNAME]}
+
+    if not self.__terminate_instances(instance_ids, conn, status_filters, 2):
+      self.handle_failure("ERROR: could not terminate instances: {}"
+                          .format(" ".join(instance_ids)))
+
     logger.info("Removing terminated instances: " + ' '.join(instance_ids))
-    conn.terminate_instances(instance_ids)
+    self.__terminate_instances(instance_ids, conn, status_filters)
 
+  def __terminate_instances(self, instance_ids, conn, status_filters, max_attempts=1):
+    """
+    Private terminate_instances that retries boto.connection.terminate_instances():
 
-  def wait_for_status_change(self, parameters, conn, state_requested,
-                             max_wait_time=60,poll_interval=10):
-    """ After we have sent a signal to the cloud infrastructure to change the state
-      of the instances (unsually from runnning to either stoppped or
-      terminated), wait for the status to change.  If all the instances change
-      successfully, return True, if not return False.
+    Bound by max_attempts:
+        State transition failures
+    *Not bound by max_attempts:*
+        If an Instance ID is not found by EC2, retry with a subset of Ids.
+        Assumption is that the missing id is terminated/deleted.
 
     Args:
-      parameters: A dictionary of parameters.
+       instance_ids: Set of instance ids to terminate
+       conn: EC2 Connection
+       status_filters: Dictionary of EC2 filters to use
+       max_attempts: Number of terminate_instances() calls to attempt
+    Returns:
+       True if instances were terminated
+       False if we were unable to terminate the instances.
+    """
+    attempts = 0
+    while attempts < max_attempts:
+      try:
+        logger.debug("Terminating instances: {} attempt: {} of {}"
+                     .format(' '.join(instance_ids), attempts, max_attempts))
+        conn.terminate_instances(list(instance_ids))
+        if self.wait_for_status_change(instance_ids, conn, status_filters,
+                                       max_wait_time=120):
+          return True
+        else:
+          attempts += 1
+
+      except boto.exception.EC2ResponseError as resp_error:
+        if resp_error.error_code == 'InvalidInstanceID.NotFound':
+          instance_ids.difference_update(re.findall('i-[a-zA-Z0-9]+',
+                                         resp_error.error_message))
+          logger.debug("New instance_ids: {}".format(' '.join(instance_ids)))
+          if len(instance_ids) == 0:
+            return True
+
+        # make another attempt
```

Comment:
For any other ec2 error we may want to have a small delay before retrying and should increment `attempts` to ensure we cannot repeat the same request indefinitely.